service/models.py

```diff
@@ -140,21 +140,30 @@ def find_by_query_string(cls, args):
         return data.all()
 
     @classmethod
-    def apply_best_promo(cls, product):
+    def apply_best_promo(cls, product_id, pricing):
         """ Find a Promotion by query string """
-        logger.info(" Finding best promotion for the product %s ...", product)
+        logger.info(" Finding best promotion for the product %s ...", product_id)
         promos = cls.query.filter(cls.start_date <= datetime.now()).filter(cls.end_date >= datetime.now())
+                
+        product_promos = promos.filter(cls.products.any(id = product_id))
+        site_wide_promos = promos.filter(cls.is_site_wide == True)
+        logger.info("  Available site wide promos: " + str(site_wide_promos.all()))
+        logger.info("  Available promos for this product: " + str(product_promos.all()))
+
         best_promo, best_discount = None, 0
-        for p in promos:
+        for p in site_wide_promos.all() + product_promos.all():
             if p.promo_type == PromoType.DISCOUNT:
                 if p.amount > best_discount:
                     best_promo, best_discount = p, p.amount
             elif p.promo_type == PromoType.BOGO and best_discount < 50:
                     best_promo, best_discount = p, 50
```

Comment:
so here we are just giving 50% off if the product has a BOGO associated with it? 
I thought we store an amount for BOGO promotions also and that could be used instead of 50. Also we are just assuming that the cart has at least 2 of those products? BOGO is confusing... I kinda wish we did not include it lol

Reply:
Indeed, I had the same issue. Right now, the two things the API considers is ``` product = product_price ``` where **product_price** is the price of a single product. For BOGO, we might need to also consider the number of products or place a flag that it only applies for even number of products. For now, I assumed it as a 50% discount but yeah it can be better approached.

Reply:
this seems good enough for now